File: Coding-Patterns/Coding-Patterns-Python/03_graphs/course_schedule.py
# LeetCode 207 - Course Schedule
# Difficulty: Medium
# Pattern: Topological sort / cycle detection (Kahn's BFS)
#
# Problem: There are n courses. Given prerequisites [a, b] meaning
# you must take course b before a, return True if you can finish all courses.
#
# Approach: Build directed graph + in-degree array.
# BFS from all nodes with in_degree=0. If we process all n nodes, no cycle exists.
#
# Time: O(V + E)  Space: O(V + E)

from typing import List
from collections import deque

def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    in_degree = [0] * num_courses
    adj = [[] for _ in range(num_courses)]

    for course, prereq in prerequisites:
        adj[prereq].append(course)
        in_degree[course] += 1

    queue = deque(i for i in range(num_courses) if in_degree[i] == 0)
    completed = 0

    while queue:
        course = queue.popleft()
        completed += 1
        for next_course in adj[course]:
            in_degree[next_course] -= 1
            if in_degree[next_course] == 0:
                queue.append(next_course)

    return completed == num_courses


# ---- Tests ----
if __name__ == "__main__":
    assert can_finish(2, [[1, 0]]) == True         # 0 -> 1, no cycle
    assert can_finish(2, [[1, 0], [0, 1]]) == False # cycle
    assert can_finish(3, [[1,0],[2,1],[3,2] if False else [2,0]]) == True
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/03_graphs/CourseSchedule.cs
// LeetCode 207 - Course Schedule
// Difficulty: Medium
// Pattern: Topological sort (Kahn's BFS)
//
// Time: O(V + E)  Space: O(V + E)

namespace CodingPatterns.Graphs;

public class CourseSchedule
{
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        var inDegree = new int[numCourses];
        var adj = new List<int>[numCourses];
        for (int i = 0; i < numCourses; i++) adj[i] = new List<int>();

        foreach (var p in prerequisites)
        {
            adj[p[1]].Add(p[0]);
            inDegree[p[0]]++;
        }

        var queue = new Queue<int>();
        for (int i = 0; i < numCourses; i++)
            if (inDegree[i] == 0) queue.Enqueue(i);

        int completed = 0;
        while (queue.Count > 0)
        {
            int course = queue.Dequeue();
            completed++;
            foreach (var next in adj[course])
                if (--inDegree[next] == 0) queue.Enqueue(next);
        }
        return completed == numCourses;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new CourseSchedule();

        Console.WriteLine(sol.CanFinish(2, new[] { new[] { 1, 0 } }));                        // True
        Console.WriteLine(sol.CanFinish(2, new[] { new[] { 1, 0 }, new[] { 0, 1 } }));        // False (cycle)
        Console.WriteLine(sol.CanFinish(3, Array.Empty<int[]>()));                            // True
        Console.WriteLine(sol.CanFinish(4, new[] { new[] { 1, 0 }, new[] { 2, 1 }, new[] { 3, 2 } })); // True
    }
}
